pylearn2/utils/iteration.py

```diff
@@ -775,15 +777,19 @@ def __init__(self, dataset, subset_iterator, data_specs=None,
             sub_spaces = space.components
         assert len(source) == len(sub_spaces)
 
-        self._raw_data = ()
-        for s in source:
-            try:
-                self._raw_data += (all_data[dataset_source.index(s)],)
-            except ValueError as e:
-                msg = str(e) + '\nThe dataset does not provide '\
-                               'a source with name: '+s+'.'
-                reraise_as(ValueError(msg))
-
+        # If `dataset` is incompatible with the new interface, fall back to the
+        # old interface
+        if not hasattr(self._dataset, 'get'):
+            warnings.warn("dataset is using the old iterator interface which "
+                          "is deprecated and will become officially "
+                          "unsupported as of July 28, 2014. The dataset "
+                          "should implement a `get` method respecting the new "
+                          "interface.")
+            all_data = self._dataset.get_data()
+            if not isinstance(all_data, tuple):
+                all_data = (all_data,)
+            self._raw_data = tuple(all_data[dataset_source.index(s)]
+                                   for s in source)
```

Comment:
Can you re-add here the code that checks if `s` is indeed in `dataset_source`, and updates the exception with a more meaningful error message?
This should fix the failing test.